subsys/bluetooth/controller/ll_sw/ull_adv.c

```diff
@@ -1499,6 +1496,12 @@ void ull_adv_done(struct node_rx_event_done *done)
 
 static int init_reset(void)
 {
+#if defined(CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL)
+	for (u16_t handle = 0U; handle < BT_CTLR_ADV_MAX; handle++) {
```

Comment:
O(n) power up CPU time impact in comparison to just in time initialization.

Reply:
Instead have a file static variable that is initialized to `RADIO_TXP_DEFAULT` ? which is then used in ll_adv_enable? This variable can then also be filled using settings subsystem.

Reply:
That would require an interface between `ll_tx_power.c` and `ull_adv.c`/`ull_scan.c`.
Is that worth avoiding a loop that sets a single int in an array of structs? 
Especially given that the the standard case of `BT_CTLR_ADV_MAX=1` the loop will just be optimized away.

Reply:
It also means that you couldn't set different powers for different scan/adv sets, unless you are suggesting declaring an array of TX powers.

Reply:
> Is that worth avoiding a single loop at startup that sets a single int in an array of structs?
With advertising extensions support, this is not going to be a single loop. 

> It also means that you couldn't set different powers for different scan/adv sets, unless you are suggesting declaring an array of TX powers.

`CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL` is proprietary feature related to changing tx power of an active state/role. Your requirement is to preserve tx power value in legacy advertising and legacy scanning which is restricted to one instance each. I suggest adding new proprietary feature specific to preserving tx power of single instance. With advertising extensions the instance handles are dynamic and it is a different problem to have persistent values for changing instance allocations.

Reply:
I think the key point in your comment is 'active' role. If it is indeed for changing the state of active roles, then I suppose resetting the power each time the role goes inactive/active makes sense.

That is not how I read the Kconfig symbol however:
```
config BT_CTLR_TX_PWR_DYNAMIC_CONTROL
	bool "Tx Power Dynamic Control"
	depends on BT_HCI_VS_EXT
	help
	  Enable dynamic control of Tx power per role/connection.
	  Provides HCI VS commands to set and get the current Tx
	  power on an individual role/connection basis.
```
After reading the description, I would expect setting the TX power for the advertising role to keep that value until told to change again, not to revert to the default just because the advertising set has been stopped and restarted.

I believe this problem effects legacy and extended advertising equally, as they both go through `ll_adv_enable`.
Therefore even under the following scenario the behavior would not be as I would expect it.
```
bt_le_ext_adv_create(..., ..., &ext_adv);
hci_cmd_set_tx_power(ADV, ext_adv->handle, +8dBm);
bt_le_ext_adv_start();
```
Would still result in the default power being used, not +8dBm.

Reply:
To avoid the boot loop, I could instead add a `tx_pwr_set` field to `struct lll_adv` which is used to choose either `RADIO_TXP_DEFAULT` or `tx_pwr_lvl` in the spot where I have deleted code.
Is that more acceptable @cvinayak ?

Reply:
Not acceptable! :-)

It seems the advertising extensions API is missing the `Advertising_TX_Power` parameter that should be exposed to the applications. You do not need any proprietary implementation for initial tx power level value.

The value is hard coded to, no preference:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L7817

The correct solution is to propagate the tx power using Bluetooth extended advertising API to the `ll_adv_params_set` function in the controller. The controller can then initialize the `tx_pwr_lvl` in `ll_adv_params_set` instead of currently in `ll_adv_enable`. 

@jhedberg @joerchan any objections?

Reply:
Looks like a mix of vendor specific and standard HCI by doing that in that case.
Adding TX power control to the advertising API is possible as an enhancement, but it would only be for the extended advertising API in that case.

Reply:
@joerchan  For the legacy advertising and scanning, they are single instances each and two file static variable can be used.

Reply:
Does https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/ll_sw/ll_tx_pwr.c#L57 look like this in that case then?
```
#if defined(CONFIG_BT_CTLR_ADV_EXT)
    adv->lll.tx_pwr_lvl = *tx_pwr_lvl;
#else
    ull_adv_set_file_static_variable(*tx_pwr_level);
#endif /* CONFIG_BT_CTLR_ADV_EXT */
```

Reply:
No. The controller change will be in ull_adv.c only for advertising tx power.

Reply:
Then how does the file static variable get set from `ll_tx_pwr_lvl_set` which is handling `BT_HCI_OP_VS_WRITE_TX_POWER_LEVEL`?

Reply:
The file static variable is nothing but `ll_adv[0].lll.tx_pwr_lvl`. 

in the `init_reset`
```
#if defined(CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL) && !defined(CONFIG_BT_CTLR_ADV_EXT)
    lll_adv[0].lll.tx_pwr_lvl = RADIO_TXP_DEFAULT;
#endif 
```

And of course, your change to remove `lll->tx_pwr_lvl = RADIO_TXP_DEFAULT;` in `ll_adv_enable`.

For the case of advertising extensions support, with enhancements to host API, `lll->tx_pwr_lvl = lll_radio_tx_pwr_floor(*tx_pwr);` has to be added in a suitable place in `ll_adv_params_set`

Reply:
@cvinayak Thanks for clearing that up, file static variable in my mind meant:
```
static struct ll_adv_set ll_adv[BT_CTLR_ADV_MAX];
static int8_t ll_tx_power = RADIO_TXP_DEFAULT;
```